File: domain/bkt.py
"""Bayesian Knowledge Tracing (BKT) Engine — Tầng 2 cốt lõi.

Tính xác suất thành thạo từng kỹ năng của học sinh dựa trên lịch sử trả lời,
sau đó truy tìm gốc rễ lỗ hổng qua Knowledge Graph.

Tham số BKT cố định (docs/ai-danh-gia.md §2.1): không tự ước lượng từ dữ liệu
vì bản demo không có đủ dữ liệu. Giá trị này minh bạch và dễ giải thích.
"""
from domain.knowledge_graph import (
    get_prerequisites,
    get_skill_name,
    has_skill,
    trace_root_causes,
)

# Tham số BKT cố định
PRIOR = 0.3        # xác suất đã biết trước khi học
TRANSIT = 0.3      # xác suất học được sau mỗi lượt
GUESS = 0.2        # xác suất đoán đúng dù chưa biết
SLIP = 0.1         # xác suất làm sai dù đã biết

# Ngưỡng đánh giá trạng thái
# Lưu ý: do tham số transit luôn cộng tiến bộ mỗi lượt (kể cả khi sai), xác suất
# sau một chuỗi toàn sai sẽ không xuống dưới ~0.34. Vì vậy ngưỡng "developing"
# đặt ở 0.45 để các kỹ năng học sinh làm sai phần lớn vẫn bị đánh dấu là lỗ hổng.
THRESHOLD_MASTERED = 0.7
THRESHOLD_DEVELOPING = 0.45
ROOT_THRESHOLD = 0.5  # ngưỡng truy tìm gốc rễ


def _status(probability: float) -> str:
    if probability >= THRESHOLD_MASTERED:
        return "mastered"
    if probability >= THRESHOLD_DEVELOPING:
        return "developing"
    return "weak"


def update_probability(prior_p: float, correct: bool) -> float:
    """Cập nhật xác suất thành thạo theo 2 bước (docs §2.2).

    Bước 1: cập nhật theo bằng chứng (đúng/sai).
    Bước 2: cộng tiến bộ tự nhiên qua lượt luyện (transit).
    """
    if correct:
        # P(đúng | biết) = 1 - slip ; P(đúng | không biết) = guess
        num = prior_p * (1 - SLIP)
        den = prior_p * (1 - SLIP) + (1 - prior_p) * GUESS
        posterior = num / den if den > 0 else prior_p
    else:
        # P(sai | biết) = slip ; P(sai | không biết) = 1 - guess
        num = prior_p * SLIP
        den = prior_p * SLIP + (1 - prior_p) * (1 - GUESS)
        posterior = num / den if den > 0 else 0.0

    # Bước 2: tiến bộ tự nhiên
    updated = posterior + (1 - posterior) * TRANSIT
    return max(0.0, min(1.0, updated))


def compute_mastery(answers: list, questions: list = None) -> dict:
    """Tính mastery theo từng skill_id từ danh sách câu trả lời.

    `answers`: list dict có {question_id, correct, skill_id, ...}
    Trả về { skill_id: {probability, correct, total, skill_name, status, is_default} }
    """
    # Khởi tạo xác suất prior cho mỗi skill được đề cập
    mastery = {}
    for a in answers or []:
        sid = a.get("skill_id")
        if sid and sid not in mastery:
            mastery[sid] = {
                "probability": PRIOR,
                "correct": 0,
                "total": 0,
                "skill_name": get_skill_name(sid),
                "status": _status(PRIOR),
                "is_default": True,
            }

    for a in answers or []:
        sid = a.get("skill_id")
        if not sid or sid not in mastery:
            continue
        m = mastery[sid]
        m["is_default"] = False
        m["total"] += 1
        correct = bool(a.get("correct"))
        if correct:
            m["correct"] += 1
        m["probability"] = update_probability(m["probability"], correct)
        m["status"] = _status(m["probability"])

    return mastery


def diagnose_gaps(mastery_map: dict, skill_ids: list = None) -> list:
    """Chẩn đoán lỗ hổng + truy tìm gốc rễ (docs §2.3, §2.4).

    Trả về list các gap:
    {
      skill_id, skill_name, probability, status, severity,
      root_causes: [skill_id...], reason
    }
    """
    gaps = []
    # Mở rộng tập kỹ năng cần xét: các kỹ năng đã làm + toàn bộ tiên quyết truy vết ngược
    # (đệ quy), để cả kỹ năng "gốc" (không có trong bài làm) cũng được chẩn đoán.
    skills_to_check = set(skill_ids or list(mastery_map.keys()))

    def _expand(sid):
        for p in get_prerequisites(sid):
            if p not in skills_to_check:
                skills_to_check.add(p)
                _expand(p)

    for sid in list(skills_to_check):
        _expand(sid)

    for sid in skills_to_check:
        if not has_skill(sid):
            # Kỹ năng không có trong đồ thị -> báo lỗi rõ (§2.4)
            gaps.append({
                "skill_id": sid,
                "skill_name": get_skill_name(sid),
                "probability": mastery_map.get(sid, {}).get("probability"),
                "status": "unknown",
                "severity": "unknown",
                "root_causes": [],
                "reason": f"Kỹ năng '{sid}' không tồn tại trong đồ thị kiến thức — cần bổ sung dữ liệu.",
            })
            continue

        prob = mastery_map.get(sid, {}).get("probability", PRIOR)
        if prob >= THRESHOLD_DEVELOPING:
            continue  # kỹ năng này ổn, không phải lỗ hổng

        roots = trace_root_causes(sid, mastery_map, threshold=ROOT_THRESHOLD)
        severity = "high" if prob < 0.15 else "medium"
        reason = f"Chưa nắm vững {get_skill_name(sid)}"
        if roots:
            root_names = ", ".join(get_skill_name(r) for r in roots)
            reason += f". Gốc rễ: {root_names}."

        gaps.append({
            "skill_id": sid,
            "skill_name": get_skill_name(sid),
            "probability": round(prob, 3),
            "status": _status(prob),
            "severity": severity,
            "root_causes": roots,
            "reason": reason,
        })

    # Sắp xếp: high trước, rồi theo probability tăng dần
    gaps.sort(key=lambda g: (g["severity"] != "high", g.get("probability") or 0))
    return gaps


def run_assessment(answers: list, questions: list = None) -> dict:
    """Hàm tiện ích: chạy toàn bộ Tầng 2, trả về {mastery, gaps}."""
    mastery = compute_mastery(answers, questions)
    gaps = diagnose_gaps(mastery)
    return {"mastery": mastery, "gaps": gaps}


# ---------------------------------------------------------------------------
# Tầng 2.5: Sinh danh sách bước lộ trình CÓ CẤU TRÚC (deterministic, không LLM)
# ---------------------------------------------------------------------------
# Đây là nơi QUANH GỌI nhất: BKT + Knowledge Graph quyết định THỨ TỰ và NỘI DUNG
# từng bước. LLM chỉ nhận output của hàm này và diễn giải thành văn bản tự nhiên.

# Độ khó gợi ý theo mastery
_DIFFICULTY_MAP = {
    "high": "easy",      # gốc rễ yếu nhất -> bắt đầu từ dễ
    "medium": "medium",
}

# Thời lượng gợi ý (phút/buổi) theo severity
_DURATION_MAP = {
    "high": "20 phút",
    "medium": "15 phút",
}


def generate_learning_steps(mastery: dict, gaps: list) -> list:
    """Sinh danh sách bước lộ trình học tập có cấu trúc từ kết quả BKT.

    Trả về list[dict], mỗi dict:
    {
      "step_order": int,           # thứ tự (1-based)
      "skill_id": str,
      "skill_name": str,
      "current_mastery": float,    # probability hiện tại (0-1)
      "status": str,               # weak / developing / mastered
      "severity": str,             # high / medium
      "reason": str,               # tại sao bước này được chọn
      "suggested_difficulty": str, # easy / medium
      "estimated_duration": str,   # "15 phút" / "20 phút"
    }

    Nguyên tắc sắp xếp:
    1. Gốc rễ (root causes) có severity "high" lên trước.
    2. Sau đó đến các kỹ năng "medium" theo thứ tự tiên quyết.
    3. Không thêm kỹ năng ngoài danh sách gaps.
    """
    if not gaps:
        return []

    steps = []
    step_num = 0

    # Bước 1: các gap có severity "high" (gốc rễ)
    for g in gaps:
        if g.get("severity") != "high":
            continue
        step_num += 1
        sid = g.get("skill_id", "")
        prob = g.get("probability") or 0
        steps.append({
            "step_order": step_num,
            "skill_id": sid,
            "skill_name": g.get("skill_name", sid),
            "current_mastery": round(prob, 3),
            "status": g.get("status", _status(prob)),
            "severity": "high",
            "reason": g.get("reason", f"Gốc rễ lỗ hổng — cần ưu tiên ôn trước"),
            "suggested_difficulty": "easy",
            "estimated_duration": _DURATION_MAP["high"],
        })

    # Bước 2: các gap có severity "medium"
    for g in gaps:
        if g.get("severity") != "medium":
            continue
        step_num += 1
        sid = g.get("skill_id", "")
        prob = g.get("probability") or 0
        root_info = ""
        if g.get("root_causes"):
            root_names = [get_skill_name(r) for r in g["root_causes"]]
            root_info = f". Tiên quyết: {', '.join(root_names)}"
        steps.append({
            "step_order": step_num,
            "skill_id": sid,
            "skill_name": g.get("skill_name", sid),
            "current_mastery": round(prob, 3),
            "status": g.get("status", _status(prob)),
            "severity": "medium",
            "reason": g.get("reason", f"Cần củng cố thêm") + root_info,
            "suggested_difficulty": _DIFFICULTY_MAP["medium"],
            "estimated_duration": _DURATION_MAP["medium"],
        })

    return steps


def format_steps_for_llm(steps: list) -> str:
    """Format danh sách bước thành bảng có cấu trúc để truyền vào prompt LLM.

    Đây là INPUT DỮ LIỆU cho LLM — LLM chỉ được diễn giải nội dung này,
    không được thêm/bớt bước hay đổi thứ tự.
    """
    if not steps:
        return "(Không có bước lộ trình — học sinh đã nắm vững tất cả kỹ năng.)"

    lines = [
        "LO TRÌNH HỌC TẬP (dữ liệu từ BKT Engine — KHÔNG được sửa đổi thứ tự hoặc nội dung):",
        "",
        "| # | Kỹ năng | Mastery hiện tại | Mức ưu tiên | Độ khó gợi ý | Thời lượng | Lý do |",
        "|---|---------|-----------------|-------------|--------------|------------|-------|",
    ]
    for s in steps:
        mastery_pct = f"{round(s['current_mastery'] * 100)}%"
        lines.append(
            f"| {s['step_order']} | {s['skill_name']} ({s['skill_id']}) | {mastery_pct} | "
            f"{s['severity']} | {s['suggested_difficulty']} | {s['estimated_duration']} | "
            f"{s['reason']} |"
        )
    return "\n".join(lines)


def get_input_skill_ids(steps: list) -> list:
    """Trích xuất danh sách skill_id từ steps — dùng cho post-validation."""
    return [s["skill_id"] for s in steps]

File: frontend/src/offline/bkt.js
/**
 * Bayesian Knowledge Tracing (BKT) Engine — Port 1:1 từ domain/bkt.py + knowledge_graph.py
 * Pure math, không I/O, chạy hoàn toàn offline trên trình duyệt.
 */

// ── Knowledge Graph ──────────────────────────────────────────────────────────

const SKILLS = {
  "G3U04-VOC": { name: "Bộ phận cơ thể (eyes, nose, hands...)", grade: 3 },
  "G3U04-PAT": { name: "Touch your... / These are my...", grade: 3 },
  "G3U09-VOC": { name: "Màu sắc (red, blue, green...)", grade: 3 },
  "G3U09-PAT": { name: "What colour is it? - It's...", grade: 3 },
  "G3U11-VOC": { name: "Gia đình (father, mother, brother...)", grade: 3 },
  "G3U11-PAT": { name: "Who's that? - He's my father", grade: 3 },
  "G3U14-VOC": { name: "Đồ vật phòng ngủ (bed, lamp, picture...)", grade: 3 },
  "G3U14-PAT": { name: "Vị trí: in / on / under / behind", grade: 3 },
  "G4U01-VOC": { name: "Bạn bè & quốc tịch", grade: 4 },
  "G4U01-PAT": { name: "Where are you from? - I'm from...", grade: 4 },
  "G4U02-VOC": { name: "Thời gian & giờ", grade: 4 },
  "G4U02-PAT": { name: "What time is it? - It's...o'clock", grade: 4 },
  "G4U03-VOC": { name: "Ngày trong tuần & thói quen", grade: 4 },
  "G4U03-PAT": { name: "What do you do on Mondays? - I...", grade: 4 },
  "G4U05-VOC": { name: "Khả năng - Things we can do", grade: 4 },
  "G4U05-PAT": { name: "Can you...? - Yes, I can / No, I can't", grade: 4 },
}

const PREREQUISITES = {
  "G3U04-PAT": ["G3U04-VOC"],
  "G3U09-VOC": ["G3U04-PAT"],
  "G3U09-PAT": ["G3U09-VOC"],
  "G3U11-VOC": ["G3U09-PAT"],
  "G3U11-PAT": ["G3U11-VOC"],
  "G3U14-VOC": ["G3U11-PAT"],
  "G3U14-PAT": ["G3U14-VOC"],
  "G4U01-VOC": ["G3U14-PAT"],
  "G4U01-PAT": ["G4U01-VOC"],
  "G4U02-VOC": ["G4U01-PAT"],
  "G4U02-PAT": ["G4U02-VOC"],
  "G4U03-VOC": ["G4U02-PAT"],
  "G4U03-PAT": ["G4U03-VOC"],
  "G4U05-VOC": ["G4U03-PAT"],
  "G4U05-PAT": ["G4U05-VOC"],
}

function get_skill_name(skill_id) {
  return (SKILLS[skill_id] || {}).name || skill_id
}

function get_prerequisites(skill_id) {
  return [...(PREREQUISITES[skill_id] || [])]
}

function has_skill(skill_id) {
  return skill_id in SKILLS
}

function trace_root_causes(skill_id, mastery_map, threshold = 0.5) {
  if (!has_skill(skill_id)) return []
  const prob = (mastery_map[skill_id] || {}).probability ?? 1.0
  if (prob >= threshold) return []

  const prereqs = get_prerequisites(skill_id)
  if (!prereqs.length) return [skill_id]

  const roots = []
  for (const p of prereqs) {
    const sub = trace_root_causes(p, mastery_map, threshold)
    if (sub.length) {
      roots.push(...sub)
    } else if (((mastery_map[p] || {}).probability ?? 1.0) < threshold) {
      roots.push(p)
    }
  }
  // dedupe preserving order
  return [...new Set(roots)]
}

// ── BKT Parameters ──────────────────────────────────────────────────────────

const PRIOR = 0.3
const TRANSIT = 0.3
const GUESS = 0.2
const SLIP = 0.1

const THRESHOLD_MASTERED = 0.7
const THRESHOLD_DEVELOPING = 0.45
const ROOT_THRESHOLD = 0.5

function _status(probability) {
  if (probability >= THRESHOLD_MASTERED) return "mastered"
  if (probability >= THRESHOLD_DEVELOPING) return "developing"
  return "weak"
}

// ── Core BKT ────────────────────────────────────────────────────────────────

function update_probability(prior_p, correct) {
  let posterior
  if (correct) {
    const num = prior_p * (1 - SLIP)
    const den = prior_p * (1 - SLIP) + (1 - prior_p) * GUESS
    posterior = den > 0 ? num / den : prior_p
  } else {
    const num = prior_p * SLIP
    const den = prior_p * SLIP + (1 - prior_p) * (1 - GUESS)
    posterior = den > 0 ? num / den : 0.0
  }
  const updated = posterior + (1 - posterior) * TRANSIT
  return Math.max(0.0, Math.min(1.0, updated))
}

function compute_mastery(answers) {
  const mastery = {}
  for (const a of answers || []) {
    const sid = a.skill_id
    if (sid && !(sid in mastery)) {
      mastery[sid] = {
        probability: PRIOR,
        correct: 0,
        total: 0,
        skill_name: get_skill_name(sid),
        status: _status(PRIOR),
        is_default: true,
      }
    }
  }
  for (const a of answers || []) {
    const sid = a.skill_id
    if (!sid || !(sid in mastery)) continue
    const m = mastery[sid]
    m.is_default = false
    m.total += 1
    const is_correct = !!a.correct
    if (is_correct) m.correct += 1
    m.probability = update_probability(m.probability, is_correct)
    m.status = _status(m.probability)
  }
  return mastery
}

function diagnose_gaps(mastery_map, skill_ids) {
  const gaps = []
  const skills_to_check = new Set(skill_ids || Object.keys(mastery_map))

  function _expand(sid) {
    for (const p of get_prerequisites(sid)) {
      if (!skills_to_check.has(p)) {
        skills_to_check.add(p)
        _expand(p)
      }
    }
  }

  for (const sid of [...skills_to_check]) _expand(sid)

  for (const sid of skills_to_check) {
    if (!has_skill(sid)) {
      gaps.push({
        skill_id: sid,
        skill_name: get_skill_name(sid),
        probability: (mastery_map[sid] || {}).probability || null,
        status: "unknown",
        severity: "unknown",
        root_causes: [],
        reason: `Kỹ năng '${sid}' không tồn tại trong đồ thị kiến thức — cần bổ sung dữ liệu.`,
      })
      continue
    }

    const prob = (mastery_map[sid] || {}).probability ?? PRIOR
    if (prob >= THRESHOLD_DEVELOPING) continue

    const roots = trace_root_causes(sid, mastery_map, ROOT_THRESHOLD)
    const severity = prob < 0.15 ? "high" : "medium"
    let reason = `Chưa nắm vững ${get_skill_name(sid)}`
    if (roots.length) {
      const root_names = roots.map(r => get_skill_name(r)).join(", ")
      reason += `. Gốc rễ: ${root_names}.`
    }

    gaps.push({
      skill_id: sid,
      skill_name: get_skill_name(sid),
      probability: Math.round(prob * 1000) / 1000,
      status: _status(prob),
      severity,
      root_causes: roots,
      reason,
    })
  }

  gaps.sort((a, b) => {
    if (a.severity !== "high" && b.severity === "high") return 1
    if (a.severity === "high" && b.severity !== "high") return -1
    return (a.probability || 0) - (b.probability || 0)
  })
  return gaps
}

function run_assessment(answers) {
  const mastery = compute_mastery(answers)
  const gaps = diagnose_gaps(mastery)
  return { mastery, gaps }
}

// ── Learning Steps (roadmap structure) ──────────────────────────────────────

const DIFFICULTY_MAP = { high: "easy", medium: "medium" }
const DURATION_MAP = { high: "20 phút", medium: "15 phút" }

function generate_learning_steps(mastery, gaps) {
  if (!gaps || !gaps.length) return []
  const steps = []
  let step_num = 0

  for (const g of gaps) {
    if (g.severity !== "high") continue
    step_num++
    const sid = g.skill_id || ""
    const prob = g.probability || 0
    steps.push({
      step_order: step_num,
      skill_id: sid,
      skill_name: g.skill_name || sid,
      current_mastery: Math.round(prob * 1000) / 1000,
      status: g.status || _status(prob),
      severity: "high",
      reason: g.reason || "Gốc rễ lỗ hổng — cần ưu tiên ôn trước",
      suggested_difficulty: "easy",
      estimated_duration: DURATION_MAP.high,
    })
  }

  for (const g of gaps) {
    if (g.severity !== "medium") continue
    step_num++
    const sid = g.skill_id || ""
    const prob = g.probability || 0
    let root_info = ""
    if (g.root_causes && g.root_causes.length) {
      const root_names = g.root_causes.map(r => get_skill_name(r))
      root_info = `. Tiên quyết: ${root_names.join(", ")}`
    }
    steps.push({
      step_order: step_num,
      skill_id: sid,
      skill_name: g.skill_name || sid,
      current_mastery: Math.round(prob * 1000) / 1000,
      status: g.status || _status(prob),
      severity: "medium",
      reason: (g.reason || "Cần củng cố thêm") + root_info,
      suggested_difficulty: DIFFICULTY_MAP.medium,
      estimated_duration: DURATION_MAP.medium,
    })
  }

  return steps
}

// ── Offline Training Plan Fallback (from plan_tool.py:193) ─────────────────

function generate_offline_plan(mastery, gaps, student_name) {
  const steps = generate_learning_steps(mastery, gaps)
  if (!steps.length) {
    return {
      summary: `${student_name || "Em"} đã nắm vững tất cả kỹ năng!`,
      steps: [],
      closing: "Làm tốt lắm! Hãy tiếp tục luyện tập đều đặn.",
    }
  }

  const plan_steps = steps.map(s => ({
    step_order: s.step_order,
    skill_name: s.skill_name,
    encouragement: `Em hãy luyện kỹ năng '${s.skill_name}' nhé!`,
    practice_tip: `Làm 5-10 bài tập về '${s.skill_name}' với độ khó '${s.suggested_difficulty}'.`,
    home_tip: `Dành ${s.estimated_duration} mỗi ngày để ôn '${s.skill_name}'.`,
  }))

  return {
    summary: `${student_name || "Em"} cần ôn ${steps.length} kỹ năng theo lộ trình BKT.`,
    steps: plan_steps,
    closing: "Em làm tốt lắm! Hãy kiên trì mỗi ngày một chút!",
  }
}

export {
  SKILLS,
  PREREQUISITES,
  get_skill_name,
  get_prerequisites,
  has_skill,
  trace_root_causes,
  update_probability,
  compute_mastery,
  diagnose_gaps,
  run_assessment,
  generate_learning_steps,
  generate_offline_plan,
}
